lib/eco/plugins/pythonindentation.py

```diff
@@ -0,0 +1,305 @@
+            next_ws = self.get_whitespace(next_r)
+            if next_ws is None:
+                next_r = next_r.next_term
+                continue
+            needed, newindent = self.get_indentation_tokens_and_indent(newindent, next_ws)
+            if not self.indents_match(next_r, needed) or next_r.indent != newindent:
+                next_r.mark_changed()
+            if next_ws < ws:
+                # if newline has smaller whitespace -> mark and break
+                break
+
+            next_r = next_r.next_term
+
+    def get_indentation_tokens_and_indent(self, indent, ws):
+        needed = []
+        newindent = []
+        if ws > indent[-1]:
+            needed.append(Node(IndentationTerminal("NEWLINE")))
+            needed.append(Node(IndentationTerminal("INDENT")))
+            newindent = indent + [ws]
+        elif ws < indent[-1]:
+            needed.append(Node(IndentationTerminal("NEWLINE")))
+            while ws < indent[-1]:
+                indent.pop()
+                needed.append(Node(IndentationTerminal("DEDENT")))
+            newindent = list(indent)
+            if ws != indent[-1]:
+                # XXX in future, just ERROR here
+                needed.append(Node(IndentationTerminal("UNBALANCED")))
+        else:
+            needed.append(Node(IndentationTerminal("NEWLINE")))
+            newindent = list(indent)
+        return needed, newindent
+
+    def indents_match(self, node, needed):
+        there = []
+        n = node.next_term
+        while isinstance(n.symbol, IndentationTerminal):
+            there.append(n)
+            n = n.next_term
+
+        if len(there) != len(needed):
+            return False
+        for i in range(len(there)):
+            if there[i].symbol != needed[i].symbol:
+                return False
+        return True
+
+    def get_whitespace(self, node):
+        if not self.is_logical_line(node):
+            return None
+
+        node = node.next_term
+        while isinstance(node.symbol, IndentationTerminal):
+            node = node.next_term
+
+        if node.lookup == "<ws>":
+            return len(node.symbol.name)
+
+        return 0
+
+    def get_last_indent(self, la):
+        return self.last_indent
+        # XXX not the most performant solution as it iterates over all elements
+        # on the stack until one has it's indent level set, which will be
+        # either a return terminal or a Nonterminal with a return somewhere in
+        # its subtrees
+        # -> replace with global variable
+        for n in reversed(self.stack):
+            if n.indent and n is not la:
+                return n.indent
+
+    def set_total_indent(self, node):
+        l = []
+        if node.children:
+            for c in node.children:
+                if c.indent:
+                    l = c.indent
+        if l:
+            node.indent = l
+
+def load(caller):
```

Comment:
So plugins load themselves in a sense which also gives the plugin author an opportunity to add constraints on when to actually load the plugin or not